docs/data/data-grid/pagination/pagination.md

```diff
@@ -120,14 +160,84 @@ React.useEffect(() => {
 <DataGrid rowCount={rowCountState} />;
 ```
 
-:::warning
-The `rowCount` prop is used in server-side pagination mode to inform the DataGrid about the total number of rows in your dataset.
-This prop is ignored when the `paginationMode` is set to `client`, i.e. when the pagination is handled on the client-side.
+{{"demo": "ServerPaginationGrid.js", "bg": "inline"}}
+
+#### Unknown row count
+
+The `hasNextPage` must not be set to `false` until there are _actually_ no records left to fetch, because when `hasNextPage` becomes `false`, the Grid considers this as the last page and tries to set the `rowCount` value to a finite value.
+
+Relying on this behavior of the Grid should be the last resort though, the preferred way is to set the `rowCount` to the actual number of rows fetched from the server as soon as the information is available since the Grid cannot guess if the rows in the last page are less than the page size, and the incorrect number might be displayed in the pagination UI.
+
+For example the if the row count is `15` and the page size is `10`, on the last page, when `hasNextPage` becomes false the Grid will set the row count to be `page (2) * pageSize (10) = (20)` showing `10-20 of 20` which is incorrect, the correct value should be `10-15 of 15`.
+
+The value of the `hasNextPage` variable might become `undefined` during loading if it's handled by some external fetching hook resulting in flickering of the pagination UI, one possible solution could be to memoize the `paginationMeta`:
+
+```tsx
+const paginationMetaRef = React.useRef<GridPaginationMeta>();
+
+const paginationMeta = React.useMemo(() => {
+  if (
+    hasNextPage !== undefined &&
+    paginationMetaRef.current?.hasNextPage !== hasNextPage
+  ) {
+    paginationMetaRef.current = { hasNextPage };
+  }
+  return paginationMetaRef.current;
+}, [hasNextPage]);
```

Comment:
This approach looks better than `useEffect` in the "Known row count" section 👍🏻
Would you mind updating it there?

Reply:
Not at all, seems a nice idea 👍